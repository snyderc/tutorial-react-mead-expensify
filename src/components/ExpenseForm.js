import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

// Use local component state to track changes to all these inputs
// After submit, do something with the information (send to Redux to add or edit expense)

export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( () => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState( () => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState( () => ({ amount }));
        }
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    // The passed in value is from a destructured object
    onFocusChange = ( { focused } ) => {
        this.setState( () => ({ calendarFocused: focused }))
    };
    onSubmit = (e) => {
        e.preventDefault(); // to avoid full-page refresh, etc.
        if (!this.state.description || !this.state.amount) {
            // Set error state equal to 'please provide desc/amt'
            this.setState( () => ({ error: 'Error: Please provide both description and amount'}));
        } else {
            // Clear the error and submit
            this.setState( () => ({ error: ''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount)*100, // format: store as cents
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                <input
                    type="text"
                    placeholder="Description"
                    autoFocus
                    className="text-input"
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    type="text"
                    placeholder="Amount"
                    className="text-input"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker 
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    className="textarea"
                    onChange={this.onNoteChange}
                >
                </textarea>
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        )
    }
}