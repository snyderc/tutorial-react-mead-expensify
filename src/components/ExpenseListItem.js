// Description, amount, createdAt

import React from 'react';
import { Link } from 'react-router-dom';

//ExpenseListItem
// What's being passed in is the partially destructured "props"
export const ExpenseListItem = ( {id, description, amount, createdAt} ) => (
    <div>
        <p>Description: <Link to={`/edit/${id}`}>{description}</Link></p>
        <p>Amount: {amount}</p>
        <p>Created at: {createdAt}</p>
    </div>
);

export default ExpenseListItem;