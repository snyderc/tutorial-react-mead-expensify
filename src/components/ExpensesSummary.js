import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';

import selectExpenses from '../selectors/expenses';
import getExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = (props) => (
    <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title">Viewing <span>{props.expenses.length}</span> expense{props.expenses.length !== 1 && 's'} totalling <span>{numeral(getExpensesTotal(props.expenses)/100).format('$0,0.00')}</span></h1>
            <div className="page-header__actions">
                <Link className="button" to="/create">Add Expense</Link>
            </div>
        </div>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

/*
Andrew Mead's solution:
const visibleExpenses = selectExpenses()
then return an object with expenseCount and expenseTotal
That way, the test case just uses a # and $, not the expense array
Also, nomenclature: Import expenses-total as selectExpensesTotal
*/

export default connect(mapStateToProps)(ExpensesSummary);