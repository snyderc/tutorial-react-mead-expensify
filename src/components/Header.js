import React from 'react';
import { NavLink } from 'react-router-dom';
import { connect } from 'react-redux';

import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
    <header>
        <h1>Expensify</h1>
        <NavLink to='/dashboard' activeClassName='is-active'>
            Dashboard
        </NavLink>
         | 
        <NavLink to='/create' activeClassName='is-active'>
            Create Expense
        </NavLink>
        <button onClick={startLogout}>Logout</button>
    </header>
)

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);

/*
<NavLink to='/' activeClassName='is-active' exact={true}>
The "exact={true}" prop is needed to go to the homepage
because otherwise to='/' will match any route
starting with that string.
*/