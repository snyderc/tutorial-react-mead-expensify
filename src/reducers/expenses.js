const expensesReducerDefaultState = [];

// Default value for expenses is empty array
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            // {id} destructures expense in the filter
            return state.filter( ({id}) => id !== action.id );
        case 'EDIT_EXPENSE':
            return state.map( (expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense,
                        ...action.updates
                    }
                } else {
                    return expense;
                }
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default:
            return state;
    }
};