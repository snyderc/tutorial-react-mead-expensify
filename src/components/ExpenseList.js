import React from 'react';
import { connect } from 'react-redux';

import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

// Making it an export so we can make unit tests
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map( (expense) => (
                    <ExpenseListItem key={expense.id} {...expense} />
                ))
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// making a HOC
// connect() returns a function
// connect()(ExpenseList) returns a HOC from the returned function
// "ConnectedExpenseList" in a way
export default connect(mapStateToProps)(ExpenseList);