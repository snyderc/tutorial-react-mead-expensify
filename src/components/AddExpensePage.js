import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startAddExpense } from '../actions/expenses';

// Changed to a class-based component
// so that onSubmit wouldn't have an inline function
// b/c that would get recalculated on every render
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <div className="page-header">
                    <div className="content-container">
                    <h1 className="page-header__title">Add Expense</h1>
                    </div>
                </div>
                <div className="content-container">
                    <ExpenseForm 
                        onSubmit={this.onSubmit}
                    />
                </div>
            </div>
        );
    }
}

// A react-redux function that makes it easier
// for us to put a spy on props.dispatch(addExpense(expense))
// b/c now we can say props.onSubmit(expense) up above
const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

// "mapStateToProps" not needed here, so passing "undefined" to connect()
export default connect(undefined, mapDispatchToProps)(AddExpensePage);