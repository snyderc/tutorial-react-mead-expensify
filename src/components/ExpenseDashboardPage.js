import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
    <div>
        <p>This is from my dashboard component</p>
        <ExpenseListFilters />
        <ExpenseList />
    </div>
);

export default ExpenseDashboardPage;