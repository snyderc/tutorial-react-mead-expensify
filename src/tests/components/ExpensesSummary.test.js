import React from 'react';
import { shallow } from 'enzyme';

import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test('should render message viewing multiple expenses with total amount', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test('should render message viewing 1 expense with amount', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[1]]} />);
    expect(wrapper).toMatchSnapshot();
});