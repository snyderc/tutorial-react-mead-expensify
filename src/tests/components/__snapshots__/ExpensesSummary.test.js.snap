// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render message viewing 1 expense with amount 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       expense
       totalling 
      <span>
        $1,095.00
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render message viewing multiple expenses with total amount 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        3
      </span>
       expense
      s
       totalling 
      <span>
        $1,141.95
      </span>
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
